File: lab1/blend_float.py
"""
4190.308 Computer Architecture                                                          Spring 2023

Image blending (float)

This module implements a function that blends two images together (floating point version)

@author:
    Your Name <your@email.com>

@changes:
    2023/MM/DD Your Name Comment

"""



def blend(img1, img2, height, width, channels, overlay, alpha):
    """
    Alpha-blends two images of size heightxwidth. The image data must contain an alpha
    channel, i.e., 'channels' must be four

    Args:
        img1:         image 1 data (multi-level list), BGRA
        img2:         image 2 data (multi-level list), BGRA
        height:       image height
        width:        image width
        channels:     number of channels (must be 4)
        overlay:      if 1, overlay the second image over the first
                      if 0, merge the two images
        alpha:        alpha blending factor (0.0-1.0)

    Returns:
        A tuple containing the following elements:
        - blended:    blended image data (multi-level list), BGRA
        - bheight:    blended image height (=height)
        - bwidth:     blended image width (=width)
        - bchannels:  blended image channels (=channels)

    """

    if channels != 4:
        raise ValueError('Invalid number of channels')


    # TODO
    # Your work goes here

    # For now, we simply copy the input parameters into the output parameters.
    # Fix/adjust once you have implemented your solution.
    bheight = height
    bwidth = width
    bchannels = channels
    blended = []

    # For now, we simply copy the input parameters into the output parameters.
    # Fix/adjust once you have implemented your solution.
    if overlay == 0:
        beta = 1 - alpha
        for h in range(bheight):
            h_temp = []
            for w in range(bwidth):
                w_temp = []
                A1 = (img1[h][w][3]/255)
                A2 = (img2[h][w][3]/255)
                A = A1 * beta + A2 * alpha
                for c in range(3):
                    temp = (img1[h][w][c] * A1 * beta + img2[h][w][c] * A2 * alpha)/255
                    w_temp.append(int(temp*255))
                w_temp.append(int(A*255))
                h_temp.append(w_temp)
            blended.append(h_temp)

    else:
        for h in range(bheight):
            h_temp = []
            for w in range(bwidth):
                w_temp = []
                temp1 = img1[h][w][3]/255
                temp2 = img2[h][w][3]/255
                alpha_prime = temp2 * alpha
                beta_prime = 1 - alpha_prime
                for c in range(3):
                    b = img1[h][w][c]/255
                    a = img2[h][w][c]/255
                    temp = b * beta_prime + a * alpha_prime
                    w_temp.append(int(temp*255))
                w_temp.append(int(temp1*255))
                h_temp.append(w_temp)
            blended.append(h_temp)

    return blended, bheight, bwidth, bchannels

File: lab2/blend_float.c
//-------------------------------------------------------------------------------------------------
// 4190.308 Computer Architecture                                                       Spring 2023
//
/// @file
/// @brief Image blending (float)
///        This module implements a function that blends two images together filter (floating-point
///        version). Note that CPython uses the 'double' data type for floating point numbers, 
///        so one may observe small differences in the output if the 'float' data type is used.
///
/// @author Your Name <your@email.com>
///
/// @section changelog Change Log
/// 2023/MM/DD Your Name Comment
///
//-------------------------------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include "blend.h"


struct Image blend_float(struct Image img1, struct Image img2, int overlay, double alpha)
{
  if (img1.channels != 4) abort();


  // Initialize blended image
  struct Image blended = {
    .height   = img1.height,
    .width    = img1.width,
    .channels = img1.channels
  };
  blended.data = malloc(blended.height*blended.width*blended.channels*sizeof(uint8));
  if (blended.data == NULL) abort();

  float temp1, temp2;
  float a1, a2;
  int h, w, c, x, y;
  
  // TODO
  // Your work goes here

  if (overlay == 0)
  {
    float beta = (float) 1.0 - (float) alpha;
    for (h=0; h < blended.height; h++){
      for (w=0; w < blended.width; w++){
        a1 = (float)PIXEL(img1, h, w, 3)/(float)255.0;
        a2 = (float)PIXEL(img2, h, w, 3)/(float)255.0;
        temp1 = a1 * beta + a2 * (float) alpha;
        temp1 = (float)255.0*temp1;
        blended.data[INDEX(blended, h, w, 3)] = (uint8) temp1;
        for (c=0; c < 3; c++){
          temp2 = ((float)PIXEL(img1, h, w, c)/(float)255.0) * a1 * beta + ((float)PIXEL(img2, h, w, c)/(float)255.0) * a2 * (float) alpha;
          temp2 = temp2 * 255.0;
          blended.data[INDEX(blended, h, w, c)] = (uint8) temp2;
        }
      }
    }
  } 
  else 
  {
    float alpha_prime, beta_prime ;
    for (h=0; h < blended.height; h++){
      for (w=0; w < blended.width; w++){
        a1 = (float)PIXEL(img1, h, w, 3)/(float)255.0;
        a2 = (float)PIXEL(img2, h, w, 3)/(float)255.0;
        alpha_prime = a2 * (float) alpha;
        beta_prime = (float) 1.0 - alpha_prime;
        
        for (c=0; c < 3; c++){
          temp2 = ((float)PIXEL(img1, h, w, c)/(float)255.0)*beta_prime + ((float)PIXEL(img2, h, w, c)/(float)255.0)*alpha_prime;
          temp2 = temp2 * 255.0;
          blended.data[INDEX(blended, h, w, c)] = (uint8)temp2;
        }
        a1 = a1 * (float)255.0;
        blended.data[INDEX(blended, h, w, 3)] = (uint8)a1;
      }
    }
  }

  return blended;
}
